pdp/expr_stringset_intersect.go

```diff
@@ -0,0 +1,83 @@
+package pdp
+
+import (
+	"fmt"
+
+	"github.com/infobloxopen/go-trees/strtree"
+)
+
+type functionSetOfStringsIntersect struct {
+	first  Expression
+	second Expression
+}
+
+func makeFunctionSetOfStringsIntersect(first, second Expression) Expression {
+	return functionSetOfStringsIntersect{
+		first:  first,
+		second: second}
+}
+
+func makeFunctionSetOfStringsIntersectAlt(args []Expression) Expression {
+	if len(args) != 2 {
+		panic(fmt.Errorf("function \"intersect\" for Set of Strings needs exactly two arguments but got %d", len(args)))
+	}
+	return makeFunctionSetOfStringsIntersect(args[0], args[1])
+}
+
+func (f functionSetOfStringsIntersect) GetResultType() Type {
+	return TypeSetOfStrings
+}
+
+func (f functionSetOfStringsIntersect) describe() string {
+	return "intersect"
+}
+
+// Calculate implements Expression interface and returns calculated value
+func (f functionSetOfStringsIntersect) Calculate(ctx *Context) (AttributeValue, error) {
+	firstSet, err := ctx.calculateSetOfStringsExpression(f.first)
+	if err != nil {
+		return UndefinedValue, bindError(bindError(err, "first argument"), f.describe())
+	}
+	secondSet, err := ctx.calculateSetOfStringsExpression(f.second)
+	if err != nil {
+		return UndefinedValue, bindError(bindError(err, "second argument"), f.describe())
+	}
+
+	res := strtree.NewTree()
+	first := firstSet.Enumerate()
```

Comment:
Sorry, for late comment to the code. You don't need to enumerate both trees. You need to enumerate one and check other:
```go
res := strtree.NewTree()
for p := range firstSet.Enumerate() {
	if _, ok := secondSet.Get(p.Key); ok {
		res.InplaceInsert(p.Key, p.Value)
	}
}
```

It's a bit slower like N*log(M) instead of max(N, M) but I don't think that it's critical to the moment.

Reply:
Here, you can process cases when one of arguments is list of strings similarly.

Reply:
One more thing about set of strings. Themis expects value for strtree key-value pair to be integer and tries to order strings in a set by the value where it needs ordered set of strings.